src/avl.py

```diff
@@ -0,0 +1,468 @@
+import json
+import logging
+
+from hypothesis import given
+import hypothesis.strategies as st
+
+
+
+class AVLTree():
+    """The interface to an AVL tree.
+
+    The tree itself is actually represented by AVLTreeNode instances, but because of
+    needing to be able to change which node is the root, we need a container class.
+
+    Args:
+      value: an initial value to insert into the tree.
+    """
+
+    def __init__(self, value=None):
+        self._root = AVLTreeNode(self)
+        if value is not None:
+            self._root.insert(value)
+
+        self.log = []
+
+    def insert(self, value):
```

Comment:
`__setitem__` seems to do the wrong thing. we're not setting the value of a thing with respect to a position/key, but rather just adding a value. note that python sets don't use setitem either.